### PR TITLE
ServiceAccountIssuerDiscovery: Add user facing documentation

### PR DESCRIPTION
**Pull request by mtaufen**
Extends the documentation on configuring service accounts to include
details on the ServiceAccountIssuerDiscovery feature.

/cc @VineethReddy02 @liggitt @mikedanese

### CODE REVIEW

**Comment by sftim** [comment]
This looks like it's pending approval for v1.18 release? If so, @mtaufen, you might want to add a /hold

File: content/en/docs/tasks/configure-pod-container/configure-service-account.md [hunk 1]
```diff
@@ -296,6 +296,59 @@ token available to the pod at a configurable file path, and refresh the token as
 
 The application is responsible for reloading the token when it rotates. Periodic reloading (e.g. once every 5 minutes) is sufficient for most usecases.
 
+## Service Account Issuer Discovery
+
+{{< feature-state for_k8s_version="v1.18" state="alpha" >}}
+
+{{< note >}}
+The Service Account Issuer Discovery feature is __alpha__ in 1.18 and enabled by
+setting the `ServiceAccountIssuerDiscovery` feature gate to `true` and then
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/configure-service-account.md]:
```suggestion
enabling the `ServiceAccountIssuerDiscovery` [feature gate](/docs/reference/command-line-tools-reference/feature-and then
```

File: content/en/docs/tasks/configure-pod-container/configure-service-account.md [hunk 2]
```diff
@@ -296,6 +296,59 @@ token available to the pod at a configurable file path, and refresh the token as
 
 The application is responsible for reloading the token when it rotates. Periodic reloading (e.g. once every 5 minutes) is sufficient for most usecases.
 
+## Service Account Issuer Discovery
+
+{{< feature-state for_k8s_version="v1.18" state="alpha" >}}
+
+{{< note >}}
+The Service Account Issuer Discovery feature is __alpha__ in 1.18 and enabled by
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/configure-service-account.md]: This is repeating the feature-state line. I would tell readers how to enable the feature gate (what component?) and omit the repetition.

File: content/en/docs/tasks/configure-pod-container/configure-service-account.md [hunk 3]
```diff
@@ -296,6 +296,59 @@ token available to the pod at a configurable file path, and refresh the token as
...
+setting the `ServiceAccountIssuerDiscovery` feature gate to `true` and then
+enabling the Service Account Token Projection feature as described above.
+
+Note that the issuer URL must comply with the
+[OIDC Discovery Spec](https://openid.net/specs/openid-connect-discovery-1_0.html). In
+practice, this means it must use the `https` scheme, and should serve an OpenID
+provider configuration at `{service-account-issuer}/.well-known/openid-configuration`.
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/configure-service-account.md]: I'd start the ` {{< note >}}` here (move the advice about enabling the feature gate to be before the note block).

File: content/en/docs/tasks/configure-pod-container/configure-service-account.md [hunk 4]
```diff
@@ -296,6 +296,59 @@ token available to the pod at a configurable file path, and refresh the token as
...
+
+If the URL does not comply, the `ServiceAccountIssuerDiscovery` endpoints will
+not be registered, even if the feature is enabled.
+{{< /note >}}
+
+The Service Account Issuer Discovery feature enables federation of Kubernetes
+service account tokens issued by a cluster (the _identity provider_) with
+external systems (_relying parties_).
+
+When enabled, the Kubernetes API server will serve an OpenID Provider
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/configure-service-account.md]:
```suggestion
When enabled, the Kubernetes API server provides a public OpenID Provider
```

File: content/en/docs/tasks/configure-pod-container/configure-service-account.md [hunk 5]
```diff
@@ -296,6 +296,59 @@ token available to the pod at a configurable file path, and refresh the token as
...
+Configuration document at `/.well-known/openid-configuration` and the associated
+JSON Web Key Set (JWKS) at `/openid/v1/jwks`. The OpenID Provider Configuration
+is sometimes referred to as the _discovery document_.
+
+{{< note >}}
+The responses served at `/.well-known/openid-configuration` and
+`/openid/v1/jwks` are designed to be OIDC compatible, but not strictly OIDC
+compliant. Today they contain only the parameters necessary to perform
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/configure-service-account.md]:
```suggestion
compliant. Those documents contain only the parameters necessary to perform
```
(“today” looks a bit like a statement about the future; see https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future)

I would definitely remove:
```markdown
Additional, parameters may be
added if it is determined that doing so improves compatibility with common
client implementations used by relying parties.
```

That kind of detail belongs in the KEP.

File: content/en/docs/tasks/configure-pod-container/configure-service-account.md [hunk 6]
```diff
@@ -296,6 +296,59 @@ token available to the pod at a configurable file path, and refresh the token as
...
+validation of Kubernetes service account tokens. Additional, parameters may be
+added if it is determined that doing so improves compatibility with common
+client implementations used by relying parties.
+{{< /note >}}
+
+The JWKS contains public keys that can be used to validate the Kubernetes
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/configure-service-account.md]:
```suggestion
The JWKS response contains public keys that can a relying party can use to validate the Kubernetes
```

File: content/en/docs/tasks/configure-pod-container/configure-service-account.md [hunk 7]
```diff
@@ -296,6 +296,59 @@ token available to the pod at a configurable file path, and refresh the token as
...
+available by users or service providers. In these cases, it is possible to
+override the `jwks_uri` in the OpenID Provider Configuration so that it points
+to the public endpoint, rather than the API server's address, by passing the
+`--service-account-jwks-uri` flag to the API server. Like the issuer URL, the
+JWKS URI is required to use the `https` scheme.
+
+For additional technical details, see the
+[Service Account Signing Key Retrieval KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/20190730-oidc-discovery.md)
+and
+[OIDC Discovery Spec](https://openid.net/specs/openid-connect-discovery-1_0.html).
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/configure-service-account.md]: These could (I think should) move into the `whatsnext` capture.

**Comment by mikedanese** [comment]
For tech review.

/lgtm

**Comment by mikedanese** [comment]
For tech review.

/lgtm

**Comment by mtaufen** [comment]
> This looks like it's pending approval for v1.18 release? If so, @mtaufen, you might want to add a /hold

@sftim The code is merged for 1.18 already. Unless you meant something else?

File: content/en/docs/tasks/configure-pod-container/configure-service-account.md (hunk 1 above)
Comment (mtaufen) [reply content/en/docs/tasks/configure-pod-container/configure-service-account.md]: Done.

File: content/en/docs/tasks/configure-pod-container/configure-service-account.md (hunk 2 above)
Comment (mtaufen) [reply content/en/docs/tasks/configure-pod-container/configure-service-account.md]: Done.

File: content/en/docs/tasks/configure-pod-container/configure-service-account.md (hunk 3 above)
Comment (mtaufen) [reply content/en/docs/tasks/configure-pod-container/configure-service-account.md]: Done.

File: content/en/docs/tasks/configure-pod-container/configure-service-account.md (hunk 4 above)
Comment (mtaufen) [reply content/en/docs/tasks/configure-pod-container/configure-service-account.md]: Done. Also added clarification on RBAC, thanks for the reminder with your comment about "public."

File: content/en/docs/tasks/configure-pod-container/configure-service-account.md (hunk 5 above)
Comment (mtaufen) [reply content/en/docs/tasks/configure-pod-container/configure-service-account.md]: Yup, already in the KEP. Sounds good. Done.

File: content/en/docs/tasks/configure-pod-container/configure-service-account.md (hunk 6 above)
Comment (mtaufen) [reply content/en/docs/tasks/configure-pod-container/configure-service-account.md]: Done.

File: content/en/docs/tasks/configure-pod-container/configure-service-account.md (hunk 7 above)
Comment (mtaufen) [reply content/en/docs/tasks/configure-pod-container/configure-service-account.md]: Done.

File: content/en/docs/tasks/configure-pod-container/configure-service-account.md [hunk 8]
```diff
@@ -296,9 +296,63 @@ token available to the pod at a configurable file path, and refresh the token as
...
+The Service Account Issuer Discovery feature enables federation of Kubernetes
+service account tokens issued by a cluster (the _identity provider_) with
+external systems (_relying parties_).
+
+When enabled, the Kubernetes API server provides an OpenID Provider
+Configuration document at `/.well-known/openid-configuration` and the associated
+JSON Web Key Set (JWKS) at `/openid/v1/jwks`. The OpenID Provider Configuration
+is sometimes referred to as the _discovery document_.
+
+When enabled, the cluster is also configured with a default RBAC `ClusterRole`
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/configure-service-account.md]: nit:
```suggestion
 When enabled, the cluster is also configured with a default RBAC ClusterRole
```
In the docs, API objects don't go inside backticks.

Comment (mtaufen) [reply content/en/docs/tasks/configure-pod-container/configure-service-account.md]: Done.

**Comment by mtaufen** [comment]
Filed the issue to remind us to create a new task webpage: https://github.com/kubernetes/website/issues/19496

@sftim do you need to `/approve` also?

**Comment by sftim** [comment]
@VineethReddy02 is docs lead for the v1.18 release

**Comment by VineethReddy02** [comment]
👋 Please rebase this PR on `dev-1.18` in order to avoid merge conflicts.

Feel free to `/hold cancel` after you've rebased.

/hold

File: content/en/docs/tasks/configure-pod-container/configure-service-account.md [hunk 9]
```diff
@@ -296,9 +296,63 @@ token available to the pod at a configurable file path, and refresh the token as
...
+## Service Account Issuer Discovery
+
+{{< feature-state for_k8s_version="v1.18" state="alpha" >}}
+
+The Service Account Issuer Discovery feature is enabled by enabling the
+`ServiceAccountIssuerDiscovery` [feature gate](/docs/reference/command-line-tools-reference/feature)
+and then enabling the Service Account Token Projection feature as described above.
+
+{{< note >}}
+Note that the issuer URL must comply with the
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/configure-service-account.md]:
```suggestion
The issuer URL must comply with the
```
The shortcode ` {{< note >}}` already marks this as a note.

File: content/en/docs/tasks/configure-pod-container/configure-service-account.md [hunk 10]
```diff
@@ -296,9 +296,63 @@ token available to the pod at a configurable file path, and refresh the token as
...
+to the public endpoint, rather than the API server's address, by passing the
+`--service-account-jwks-uri` flag to the API server. Like the issuer URL, the
+JWKS URI is required to use the `https` scheme.
 {{% /capture %}}
 
 {{% capture whatsnext %}}
-See also the
-[Cluster Admin Guide to Service Accounts](/docs/reference/access-authn-authz/service-accounts-admin/).
+See also:
+- [Cluster Admin Guide to Service Accounts](/docs/reference/access-authn-authz/service-accounts-admin/)
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/configure-service-account.md]: Add a blank line before this list to keep the Markdown parser happy.

File: content/en/docs/tasks/configure-pod-container/configure-service-account.md [hunk 11]
```diff
@@ -296,9 +296,63 @@ token available to the pod at a configurable file path, and refresh the token as
 
 The application is responsible for reloading the token when it rotates. Periodic reloading (e.g. once every 5 minutes) is sufficient for most usecases.
 
+## Service Account Issuer Discovery
+
+{{< feature-state for_k8s_version="v1.18" state="alpha" >}}
+
+The Service Account Issuer Discovery feature is enabled by enabling the
+`ServiceAccountIssuerDiscovery` [feature gate](/docs/reference/command-line-tools-reference/feature)
+and then enabling the Service Account Token Projection feature as described above.
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/configure-service-account.md]: nit:
```suggestion
and then enabling the Service Account Token Projection feature as described [above](#service-account-token-volume-projection).
```

**Comment by VineethReddy02** [comment]
Hi, @mtaufen As today is the deadline for docs placeholder PR to be ready for review state? Can we have all the requested changes to the PR sometime soon?
Thanks!

File: content/en/docs/tasks/configure-pod-container/configure-service-account.md (hunk 11 above)
Comment (mtaufen) [reply content/en/docs/tasks/configure-pod-container/configure-service-account.md]: Done.

File: content/en/docs/tasks/configure-pod-container/configure-service-account.md (hunk 10 above)
Comment (mtaufen) [reply content/en/docs/tasks/configure-pod-container/configure-service-account.md]: Done.

File: content/en/docs/tasks/configure-pod-container/configure-service-account.md (hunk 9 above)
Comment (mtaufen) [reply content/en/docs/tasks/configure-pod-container/configure-service-account.md]: Done.

**Comment by mtaufen** [comment]
/hold cancel

@VineethReddy02 all requested changes have been made.

**Pull request merged**